### PR TITLE
Use IndexOfAny(char, char) when possible

### PR DESCRIPTION
**Pull request by BrennanConroy**
Inspired by https://github.com/dotnet/aspnetcore/pull/39743, found more places to use the faster version of `IndexOfAny`, plus some bonus temporary string allocations removed from `NavigationManager`.

### CODE REVIEW

File: src/Components/Components/src/NavigationManager.cs [hunk 1]
```diff
@@ -232,9 +230,9 @@ public string ToBaseRelativePath(string uri)
             return uri.Substring(_baseUri.OriginalString.Length);
         }
 
-        var pathEndIndex = uri.IndexOfAny(UriPathEndChar);
-        var uriPathOnly = pathEndIndex < 0 ? uri : uri.Substring(0, pathEndIndex);
-        if ($"{uriPathOnly}/".Equals(_baseUri.OriginalString, StringComparison.Ordinal))
+        var pathEndIndex = uri.AsSpan().IndexOfAny('#', '?');
+        var uriPathOnly = pathEndIndex < 0 ? uri : uri.AsSpan().Slice(0, pathEndIndex);
```
Comment (JamesNK) [inline src/Components/Components/src/NavigationManager.cs:234]:
```suggestion
        var uriPathOnly = pathEndIndex < 0 ? uri : uri.AsSpan(0, pathEndIndex);
```

File: src/Components/Components/src/NavigationManager.cs [hunk 2]
```diff
@@ -232,9 +230,9 @@ public string ToBaseRelativePath(string uri)
...
+        if (_baseUri.OriginalString.EndsWith('/') && uriPathOnly.Equals(_baseUri.OriginalString.AsSpan().Slice(0, _baseUri.OriginalString.Length - 1), StringComparison.Ordinal))
```
Comment (JamesNK) [inline src/Components/Components/src/NavigationManager.cs:235]:
```suggestion
        if (_baseUri.OriginalString.EndsWith('/') && uriPathOnly.Equals(_baseUri.OriginalString.AsSpan(0, _baseUri.OriginalString.Length - 1), StringComparison.Ordinal))
```

File: src/Components/Components/src/NavigationManager.cs [hunk 3]
```diff
@@ -498,9 +496,9 @@ private static bool TryGetLengthOfBaseUriPrefix(Uri baseUri, string uri, out int
             return true;
         }
 
-        var pathEndIndex = uri.IndexOfAny(UriPathEndChar);
-        var uriPathOnly = pathEndIndex < 0 ? uri : uri.Substring(0, pathEndIndex);
-        if ($"{uriPathOnly}/".Equals(baseUri.OriginalString, StringComparison.Ordinal))
+        var pathEndIndex = uri.AsSpan().IndexOfAny('#', '?');
+        var uriPathOnly = pathEndIndex < 0 ? uri : uri.AsSpan().Slice(0, pathEndIndex);
```
Comment (JamesNK) [inline src/Components/Components/src/NavigationManager.cs:500]:
```suggestion
        var uriPathOnly = pathEndIndex < 0 ? uri : uri.AsSpan(0, pathEndIndex);
```

File: src/Components/Components/src/NavigationManager.cs [hunk 4]
```diff
@@ -498,9 +496,9 @@ private static bool TryGetLengthOfBaseUriPrefix(Uri baseUri, string uri, out int
...
+        if (baseUri.OriginalString.EndsWith('/') && uriPathOnly.Equals(baseUri.OriginalString.AsSpan().Slice(0, baseUri.OriginalString.Length - 1), StringComparison.Ordinal))
```
Comment (JamesNK) [inline src/Components/Components/src/NavigationManager.cs:501]:
```suggestion
        if (baseUri.OriginalString.EndsWith('/') && uriPathOnly.Equals(baseUri.OriginalString.AsSpan(0, baseUri.OriginalString.Length - 1), StringComparison.Ordinal))
```

**Review by JamesNK — APPROVED** [review]
Combine AsSpan + Slice. e.g. https://github.com/dotnet/runtime/pull/76799

**Comment by davidfowl** [comment]
Sweet!

File: src/Middleware/OutputCaching/src/OutputCacheKeyProvider.cs [hunk 5]
```diff
@@ -70,7 +68,7 @@ public string CreateStorageKey(OutputCacheContext context)
 
     public static bool ContainsDelimiters(string? value)
     {
-        return !string.IsNullOrEmpty(value) && value.IndexOfAny(KeyDelimiters) > -1;
+        return !string.IsNullOrEmpty(value) && value.AsSpan().IndexOfAny(KeyDelimiter, KeySubDelimiter) > -1;
```
Comment (gfoidl) [inline src/Middleware/OutputCaching/src/OutputCacheKeyProvider.cs:71]:
```suggestion
        return !string.IsNullOrEmpty(value) && value.AsSpan().IndexOfAny(KeyDelimiter, KeySubDelimiter) >= 0;
```
Is a tiny little bit better at cpu-level, as the comparison and the jump can be fused into one actual executed instruction.

File: src/Mvc/Mvc.Core/src/ModelBinding/PrefixContainer.cs [hunk 6]
```diff
@@ -150,7 +148,7 @@ private static void GetKeyFromNonEmptyPrefix(string prefix, string entry, IDicti
         {
             case '.':
                 // Handle an entry such as "prefix.key", "prefix.key.property" and "prefix.key[index]".
-                var delimiterPosition = entry.IndexOfAny(Delimiters, keyPosition);
+                var delimiterPosition = entry.AsSpan(keyPosition).IndexOfAny('[', '.');
                 if (delimiterPosition == -1)
```
Comment (gfoidl) [inline src/Mvc/Mvc.Core/src/ModelBinding/PrefixContainer.cs:152]:
```suggestion
                if (delimiterPosition < 0)
```

File: src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonInputFormatter.cs [hunk 7]
```diff
@@ -408,11 +408,15 @@ private static ModelMetadata GetPathMetadata(ModelMetadata metadata, string path
             else
             {
                 // At start of "property", "property." or "property[0]".
-                var endIndex = path.IndexOfAny(new[] { '.', '[' }, index);
+                var endIndex = path.AsSpan(index).IndexOfAny('.', '[');
                 if (endIndex == -1)
```
Comment (gfoidl) [inline src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonInputFormatter.cs:412]:
```suggestion
                if (endIndex < 0)
```

**Review by gfoidl — COMMENTED** [review]
Some micro-changes.

**Pull request merged**